src/impl_/coroutine.rs

```diff
@@ -39,10 +39,10 @@ fn get_ptr<T: PyClass>(obj: &Py<T>) -> *mut T {
 pub struct RefGuard<T: PyClass>(Py<T>);
 
 impl<T: PyClass> RefGuard<T> {
-    pub fn new(obj: Bound<'_, PyAny>) -> PyResult<Self> {
-        let owned = obj.downcast_into::<T>()?;
+    pub fn new(obj: &Bound<'_, PyAny>) -> PyResult<Self> {
+        let owned = obj.downcast::<T>()?;
         mem::forget(owned.try_borrow()?);
```

Comment:
Hmm, now that I see these `men::forget` calls I suspect that the changes I made to make `PyRef` and `PyRefMut` contain `Bound` mean that we're leaking a reference count here.

Probably not related to this PR. I'll pause reviewing here just so I remember to reread this in the morning.

Reply:
#3917 can fixup later.